source/stream.js

```diff
@@ -0,0 +1,62 @@
+import {isReadableStream} from 'is-stream';
+
+export const getAsyncIterable = stream => {
+	if (isReadableStream(stream, {checkOpen: false})) {
+		return getStreamIterable(stream);
+	}
+
+	if (stream?.[Symbol.asyncIterator] === undefined) {
+		throw new Error('The first argument must be a Readable, a ReadableStream, or an async iterable.');
+	}
+
+	return stream;
+};
+
+// The default iterable for Node.js streams does not allow for multiple readers at once, so we re-implement it
+const getStreamIterable = async function * (stream) {
+	if (nodeImports === undefined) {
+		await loadNodeImports();
+	}
+
+	const controller = new AbortController();
+	handleStreamEnd(stream, controller);
+
+	try {
+		for await (const [chunk] of nodeImports.events.on(stream, 'data', {
+			signal: controller.signal,
+			highWatermark: stream.readableHighWaterMark,
+		})) {
+			yield chunk;
+		}
```

Comment:
`on()` changed with Node 20.0.0, more specifically [this PR](https://github.com/nodejs/node/pull/41276).

That PR fixes many subtle bugs, so it took some effort to make this work on both Node 18 and 20, but this should work now.

Reply:
Maybe target Node.js 20? I think we should target Node.js 20 in Execa anyway. Node.js 18 goes out of maintenance in April, which is not far away.

Reply:
Unfortunately, I believe Node 18 is in maintenance one more year: [until April 2025](https://github.com/nodejs/Release).

I also initially thought it was out of maintenance next month. Maybe we got this impression because they unsupported Node 16 [earlier than usual](https://nodejs.org/en/blog/announcements/nodejs16-eol/) (was in September 2023 instead of April 2024) due to some security issue.

Reply:
Ugh. Yeah, I totally misread that...

Reply:
> so it took some effort to make this work on both Node 18 and 20, but this should work now.

Maybe worth adding a TODO comment about how it could be simplified when Node.js 20 can be targeted.

Reply:
I was initially using the `iterable` returned by `once()` directly, and calling `iterator.return()`/`iterator.throw()`. 

To support both Node 18 and 20, I switched to using the `signal` option, which did not change between Node 18 and 20. It turns out this is actually a better option for any Node version.

So the code can remain the same even after removing support for Node 18.